### PR TITLE
API-16163 Added HasReferences interface to all datamart models 

### PR DESCRIPTION
**Pull request by taylorrassmannlibertyits**
Added HasReferences interface to all datamart models and rippled through controllers.

### CODE REVIEW

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/allergyintolerance/DatamartAllergyIntolerance.java [hunk 1]
```diff
@@ -68,6 +83,13 @@ public Optional<DatamartReference> recorder() {
     return recorder;
   }
 
+  @Override
+  public Stream<DatamartReference> references() {
+    return Stream.concat(
+        Stream.of(recorder().orElse(null), patient()),
+        notes().stream().map(n -> n.practitioner().orElse(null)));
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/allergyintolerance/DatamartAllergyIntolerance.java:90]: As a general note, these method should filter the null elements from the stream

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/DatamartPractitionerRole.java [hunk 2]
```diff
@@ -86,6 +89,16 @@ public Optional<DatamartReference> practitioner() {
     return practitioner;
   }
 
+  @Override
+  public Stream<DatamartReference> references() {
+    return Stream.of(
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/DatamartPractitionerRole.java:94]: Prefer `Stream.concat(Stream.of(...), location().stream())` here to be consistent with the other resources

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/observation/DatamartObservation.java [hunk 3]
```diff
@@ -138,6 +142,13 @@ public Optional<ReferenceRange> referenceRange() {
     return referenceRange;
   }
 
+  @Override
+  public Stream<DatamartReference> references() {
+    return Stream.concat(
+            Stream.of(subject().orElse(null), encounter().orElse(null)), performer().stream())
```
[the remark this answers is not in the document]
Comment (taylorrassmannlibertyits) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/observation/DatamartObservation.java:148]: Good catch, this was not being done in replaceReferences, aka copy-pasta error on my part

**Comment by jblefkove-va** [comment]
`DatamartMedication`?

**Comment by taylorrassmannlibertyits** [comment]
> `DatamartMedication`?

I didn't see any references for this class

**Comment by jblefkove-va** [comment]
> > `DatamartMedication`?
> 
> I didn't see any references for this class

For consistency all the Datamart models should implement the new interface

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/encounter/DatamartEncounter.java [hunk 4]
```diff
@@ -103,6 +109,15 @@ public List<ReasonCode> reasonCode() {
     return reasonCode;
   }
 
+  @Override
+  public Stream<DatamartReference> references() {
+    return Streams.concat(
+            Stream.of(serviceProvider().orElse(null), patient(), appointment().orElse(null)),
+            Safe.stream(locations()).map(DatamartEncounter.Location::location),
+            Safe.stream(participants()).map(DatamartEncounter.Participant::participant))
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/encounter/DatamartEncounter.java:117]: these class names don't need to be qualified

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/organization/DatamartOrganization.java [hunk 5]
```diff
@@ -101,6 +104,11 @@ public Optional<String> providerId() {
     return providerId;
   }
 
+  @Override
+  public Stream<DatamartReference> references() {
+    return Stream.empty();
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/organization/DatamartOrganization.java:109]: `partOf`

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/DatamartDiagnosticReport.java [hunk 6]
```diff
@@ -74,6 +77,12 @@ public List<DatamartReference> orders() {
     return orders;
   }
 
+  @Override
+  public Stream<DatamartReference> references() {
+    return Stream.concat(Stream.of(patient(), encounter().orElse(null)), results().stream())
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/DatamartDiagnosticReport.java:82]: several potential reference fields are missing

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/device/DatamartDevice.java [hunk 7]
```diff
@@ -84,6 +87,11 @@ public Optional<String> model() {
     return model;
   }
 
+  @Override
+  public Stream<DatamartReference> references() {
+    return Stream.of(patient, location().orElse(null)).filter(Objects::nonNull);
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/device/DatamartDevice.java:92]: `patient()`

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/device/DatamartDevice.java [hunk 8]
```diff
@@ -17,7 +20,7 @@
 @Builder
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
 @AllArgsConstructor(access = AccessLevel.PRIVATE)
-public class DatamartDevice implements HasReplaceableId, AsList<DatamartDevice> {
+public class DatamartDevice implements HasReferences, HasReplaceableId, AsList<DatamartDevice> {
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/device/DatamartDevice.java:23]: while we're in here... move `objectType` and `objectVersion` to the top to be consistent with the other classes

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/organization/DatamartOrganization.java [hunk 9]
```diff
@@ -101,6 +104,11 @@ public Optional<String> providerId() {
     return providerId;
   }
 
+  @Override
+  public Stream<DatamartReference> references() {
+    return Stream.of(partOf().orElse(null));
```
Comment (jblefkove-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/organization/DatamartOrganization.java:109]: filtering

**Review by jblefkove-va — APPROVED** [review]
:shipit:

**Pull request merged**